docs/source/traits_user_manual/defining.rst

```diff
@@ -405,6 +405,57 @@ the table.
 |                  | *allow_none* = False, *adapt* = 'yes', \*\*\ *metadata*])|
 +------------------+----------------------------------------------------------+
 
+.. index:: Instance trait
+
+.. _instance:
+
+Instance
+::::::::
+One of the most fundamental and useful predefined trait types is Instance.
+Instance trait values are an instance of a particular class or its subclasses,
+as specified by the **klass** argument. **klass** can be either an instance of
+a class or a class itself (note this applies to all python classes, not
+necessarily just ``HasTraits`` subclasses). Further, **klass** can also be an
+interface, in which case the trait only accepts values implementing that
+interface. For more details on this particular usecase, see :ref:`interfaces`.
+
+If **klass** is an instance or if it is a class and **args** and **kw** are not
+specified, the default value is None. Otherwise, the default value is
+obtained by calling the callable **factory** argument (or **klass** if
+**factory** is None) with **args** and **kw**. For example::
+
+    # instance_trait_defaults.py --- Example of Instance trait default values
+    from traits.api import HasTraits, Instance
+
+    class Parent(HasTraits):
+        pass
+
+    class Child(HasTraits):
+        # default value is None
+        father = Instance(Parent())
+        # default value is also None
+        grandfather = Instance(Parent)
+        # default value is Parent()
+        mother = Instance(Parent, args=())
+
+In the first case, although we instatiate a ``Parent`` object, that instance is
+not used as the default to prevent different instances of the ``Child`` class
+from sharing the same instance of ``Parent`` for their ``father`` trait by
+default. In the last case, the default ``Parent`` instance is not immediately
+created, but rather is lazily instantiated when the trait is first accessed.
+The default ``Parent`` will also be instantiated if the trait is assigned to
+and there is a change handler defined on the trait (to detect changes from the
+default value). For more details on change handlers and trait notification see
+:ref:`observe-notification`.
+
+Somewhat surprisingly, ``mother = Instance(Parent, ())`` will also yield a
+default value of ``Parent()``, even though in that case it is **factory** that
+is ``()`` not **args**.  This is a result of implementation details, and using
+this form is discouraged.  Another discouraged behavior is passing in another
```

Comment:
On `factory` being `()`, we should make it clear what the preferred way of writing the code is.  The `Instance(SomeClass, ())` is a _very_ common idiom in existing code.

Reply:
For the second, maybe change "discouraged behaviour" to "common mistake"?

Reply:
> On factory being (), we should make it clear what the preferred way of writing the code is. The Instance(SomeClass, ()) is a very common idiom in existing code.

Just for my own understanding, the reason we are advising against this is just it is less clear/explicit and might not be doing what you think it is under the hood, correct?  Basically Instance is doing some DWIM stuff and it sees factory is a tuple `()` so it sets factory to klass and args to ().  So it effectively does exactly the same thing as `Instance(SomeClass, args=())`, but sort of secretly.

In other words my real question is there isn't any actual difference in end behavior correct? / there aren't any strange and rare bugs this directly causes?

Reply:
> For the second, maybe change "discouraged behaviour" to "common mistake"?

@mdickinson is this section added in the docs enough to close #1214.  The issue mentions that we might want to consider even making it an error if we can't think of any valid uses.  We also may want to explicitly add a warning in the code not just the documentation?

Reply:
> Just for my own understanding, the reason we are advising against this is [...]

Um, actually, I don't know. Where did you get the information that we're advising against this?!

Honestly, I don't think we can or should advise against - the pattern is so common that we should probably just keep it, and I'm not aware of any actual problems that creates.

I _would_ still *recommend* passing the args by keyword instead of position (so `Instance(SomeClass, args=())` rather than `Instance(SomeClass, ())`), just for the benefit of being explicit.

But I think that's different from actually discouraging the second form. I don't want to users to get the impression that the second form is likely to be deprecated.

Reply:
> In other words my real question is there isn't any actual difference in end behavior correct?

There's nothing that I'm aware of.

> is this section added in the docs enough to close #1214

Hmm. Good question. Probably not. I'm somewhat inclined to close #1214 anyway, though - warning about `Instance(Int)` seems like something that would be more reasonable in a static checker than in Traits itself.

Reply:
> Um, actually, I don't know. Where did you get the information that we're advising against this?!

In issue #1180 the comment "Also explicitly document (or disparage) the equivalent form Instance(SomeClass, ())".  Sounds like we are more on the document rather than disparage mindset.

> Honestly, I don't think we can or should advise against - the pattern is so common that we should probably just keep it, and I'm not aware of any actual problems that creates.
> 
> I _would_ still _recommend_ passing the args by keyword instead of position (so `Instance(SomeClass, args=())` rather than `Instance(SomeClass, ())`), just for the benefit of being explicit.
> 
> But I think that's different from actually discouraging the second form. I don't want to users to get the impression that the second form is likely to be deprecated.

That make sense, I can update the relevant sentence to not say "using this form is discouraged", and just say something like "This is a result of implementation details, however the recommended way of writing this code is to explicitly pass **args** by keyword like so `mother = Instance(Parent, args=())`."

Reply:
Yes, that sounds good.